### ISSUE TITLE
`Stripe.Metadata` has incorrect type

### ISSUE THREAD

**Issue by SpencerKaiser** [stripe/stripe-node#1553]
### Describe the bug

The type for `Stripe.Metadata` is `{ [name: string]: string; }` which is incorrect; this specifies that all keys have a value of `string` when in reality, any key could be `string` OR `undefined`.

### To Reproduce

``` typescript
const { wat } = {} as Stripe.Metadata;
```

<img width="304" alt="image" src="https://user-images.githubusercontent.com/6445731/190441984-8244a355-2adb-47d6-8346-9a3fd9390e57.png">

### Expected behavior

Any valued pulled from a metadata object should have the type `string | undefined`

### Code snippets

```JavaScript
typescript
const { wat } = {} as Stripe.Metadata;
```
```


### OS

macOS

### Node version

Node v16.13.0

### Library version

stripe 9.16.0

### API version

'2020-08-27'

### Additional context

Fixing this will be a breaking change for users who consume this type in typescript

**Comment by kamil-stripe** [comment]
Hey @SpencerKaiser! Thank you for your report. I'd like to know more about your use case. Are you able to share it with us?
I am also confused with the proposed code change. You mention in "Expected behavior" section that any value should be treated as `string | undefined` but the pull request changes the type of the key.
Lastly, the pull request does not build properly. For the proposed change, the TypeScript compiler fails with following message"
```
';' expected.
Property or signature expected.
```
See [TS playground](https://www.typescriptlang.org/play?ssl=4&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgLITHAJnTyDeAUMicgNohwC2EAXMgM5hSgDmAugPz1MsisBuQgF9ChBAHsQTAsgDuuZMOQBeAsrgM0GbLjhDJ0iQBsIAOmMTWACgVgAlEKA) to reproduce.

If we adjust the code to following:
```
interface Metadata {
    [name: string | undefined]: string;
}
```
it results in 
```
An index signature parameter type must be 'string', 'number', 'symbol', or a template literal type.
```
Let me know what you think. Looking forward to your input.

**Comment by SpencerKaiser** [comment]
@kamil-stripe 👋 `noUncheckedIndexedAccess` looks relevant, but I wouldn't necessarily want to enable that just to address this one use case. And my bad on the pr 🤦‍♂️ I actually think `[name: string]: string | undefined` is what I wanted there... thoughts?

Regarding our exact use case, we wanted to provide a means for unauthenticated users to purchase home service gifts without being authenticated ([shameless plug for that functionality](https://hellochappy.com/send-gifts) 🤪). Per the docs, we aren't saving payment intents and are instead relying on `metadata` to provide context. One of the metadata values we used was `serviceId`, which is included when a user submits specifies an exact service for the gift, but that field is `optional` in our system so it will get trimmed off the request to Stripe if we don't explicitly set it to `null` instead of `undefined`. When we receive the webhook event from Stripe, we try and pull that value out of the metadata and then retrieve info about that service. Because the type was `string` instead of `string | undefined` our team member didn't catch a potential bug when we query a table by that ID.

Original code (obfuscated):
``` typescript
const { serviceId /* other values */ } = metadata;

const service = await db.findOne(Service, { id: serviceId }); 
// 💥 Runtime error; Malformed SQL query
```

Our "patch":
``` typescript
const { serviceId /* other values */ } = metadata as Partial<Stripe.Metadata>;

const service = await db.findOne(Service, { id: serviceId }); 
// 💥 Compile time error: Type 'undefined' is not assignable to type 'string'.
```

TL;DR - I think it's safer to force devs to handle optional properties on ALL metadata values because it's impossible to force strong types on it. As a result, errors should be shifted to compile time more often vs runtime.

**Comment by kamil-stripe** [comment]
@SpencerKaiser I see. Thank you for the detailed explanation and congrats on the new functionality. Let me discuss this with the broader team and get back to you.

**Comment by pakrym-stripe** [comment]
Hi, @SpencerKaiser. 

I think the `noUncheckedIndexedAccess` is the way to go here. The metadata dictionary has the same semantics as `Record<string, string>` which doesn't return `undefined` by-default .

 `noUncheckedIndexedAccess` is a feature specifically designed to  address you use case.

**Comment by SpencerKaiser** [comment]
@pakrym-stripe it's a neat flag for sure, but I still don't think the type is accurate the way it is. Given the nature of `Metadata` there will never be a guarantee that anything in it is a string; why not be explicit and say that the keys inside that object could be strings or non-existent? 

Plus for the average user, flipping that flag to `true` is gonna cause some type errors and I doubt yall want to add documentation encouraging people to turn that flag on... just seems like the safe bet here is to change the type to be "safe" for all situations.

**Comment by pakrym-stripe** [comment]
> Given the nature of Metadata there will never be a guarantee that anything in it is a string; why not be explicit and say that the keys inside that object could be strings or non-existent?

The same is true for every other use of `Record<string , X>` in TypeScript ([this type checks](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAhgLhgJQKagE4BMA80MCWYA5gDQz5HEB8MAvDAN4C+A3AFDtwDaARHLwC6AOiggAylEIkAFAEpWQA)). Stripe library just follows the stablished pattern. 

If customers want to make indexing untyped dictionaries be a strict operation there is a standard flag for that ([same example produces an error](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBAhgLhgJQKagE4BMA80MCWYA5gDQz5HEB8MAvDAN4C+A3AFDtwDaARHLwC6AOiggAylEIkAFAEpWQA)).

Can you help me understand how Metadata dictionary is different from other places that will trigger an error when you enable `noUncheckedIndexedAccess` ?

**Comment by SpencerKaiser** [comment]
The difference is that the user isn't creating that type as part of their project; it's an established component of your SDK. When I use SDKs I expect that the maintainers will have created types that prioritize **safety** over everything else and I consider the type of that attribute to be unsafe because the layman user and the pro user alike will see the indexed value to be of type `string` when in reality there is no guarantee, unless they set a config flag that they may not even be aware of (I wasn't until @kamil-stripe pointed it out! 😅).

If I were creating my own `Record<string, X>` type in my app, I'm 100% at the mercy of that type and I would either flip that flag on OR set the type to be `X | undefined` if it was important to me. With the Stripe SDK, I _**don't**_ have control over it and aside from using something like `patch-package` to make the change myself, the only other options I have are to expect my developers to be 100% certain the value will be there (which is a risk) OR enable that flag for my entire project and deal with the consequences.

If yall make the proposed change, 100% of your users will have to address the possibility that the value in metadata may be undefined, which is a _**safer**_ practice which will likely result in fewer bugs for your customers, so it seems like a win-win.

**Comment by pakrym-stripe** [comment]
I've chatted with the team and we decided to follow what's normally done in TypeScript ecosystem and not change the definition of Metadata.

Some reasons:

1. The result of `Object.fromEntries` seems like a good analog of metadata and it's type as `{[k: string]: string}`:
<img width="346" alt="image" src="https://user-images.githubusercontent.com/99349468/217961563-a73c0543-e69c-4ac7-8937-fcc6f5ff5867.png">

2. `{[k: string]: string | undefined}` allows for 
```js
{
    a: "a",
    b: undefined
}
```
which is not correct.

3. `{[k: string]: string | undefined}` makes `Object.values()` incorrectly typed as `(string|undefined)[]`
4. We didn't hear other customers confused by the type.
5. There is standard way to make the indexer typing stricter.
